### DISCUSSION TITLE
Implementing simple cursor pagination

### DISCUSSION THREAD

**Discussion by maksalees** [diesel-rs/diesel#3732] [Q&A] [answered]
I'm trying to implement a paginate method in query builder that will call `filter`, `gt` and `lt` DSL methods depending on the content of the struct.
For example, I have a structure like this and I want to call `fillter(id.gt(pagination.min_id))` if it has a `min_id`, otherwise `lt`.
```
#[derive(Deserialize)]
pub struct Pagination {
    pub max_id: Option<String>,
    pub min_id: Option<String>,
}
```

How can such a thing be implemented? Thanks

**Comment by weiznich** [comment]
There are different ways to implement such a conditional filter. You could either use [boxed queries](https://docs.diesel.rs/2.1.x/diesel/prelude/trait.QueryDsl.html#method.into_boxed) or [`BoxableExpression`](https://docs.diesel.rs/2.1.x/diesel/expression/trait.BoxableExpression.html). The linked documentation for both approaches shows examples of how to do the conditional filtering. Both approaches have their own advantages and disadvantages so one or the other might be better suited for your use case.

**Reply by maksalees** [reply]
Is it possible to make this method generic so that it works for all tables at once?

**Reply by weiznich** [reply]
There are a lot of things possible. It always depends on the amount of work you are willing to do to implement all necessary things. In this case you need to think about a way to get information what column you are filtering on. If it's just the primary key column you might use [`Table::PrimaryKey`](https://docs.diesel.rs/2.1.x/diesel/prelude/trait.Table.html#associatedtype.PrimaryKey) to construct the relevant column type. You should be able to get a table type via the implementation of `BoxeableExpression`. This will require a few lines of nasty generic code that requires some understanding of how complicated rust trait bounds are written. You likely also need to restrict that implementation to tables that have a compatible primary key column, so for example that do not have a `INTEGER` primary key column.

**Reply by maksalees** [reply]
But in cases where `max_id` and `min_id` are both None, I would like to not use pagination at all. As I understand it, it is not possible to do this with `BoxableExperssion` without duplicating code everywhere.

Also, could you elaborate on these things? Thanks

**Reply by weiznich** [reply]
> But in cases where max_id and min_id are both None, I would like to not use pagination at all. As I understand it, it is not possible to do this with BoxableExperssion without duplicating code everywhere.

Well you could just return a filter that always evaluates to `true` in that case, for example a literal `TRUE` in sql?

> Also, could you elaborate on these things? Thanks

On what exactly? Maybe try to share what you already have and which part exactly is challenging for you? (I generally do not provide complete solutions to these kinds of problem as it's not helpful for you to end up with a bunch of code you don't understand).

**Comment by maksalees** [comment, marked as answer]
I solved this by writing a macro.

```rust
#[macro_export]
macro_rules! paginate {
    ( $query:expr, $column:path, $pagination:expr ) => {{
        let query = $query;
        match $pagination {
            Pagination::MaxId(id, limit) => query.filter($column.gt(id)).limit(limit.into()),
            Pagination::MinId(id, limit) => query.filter($column.lt(id)).limit(limit.into()),
            Pagination::None(limit) => query.limit(limit.into()),
        }
    }};
}
```
